### ISSUE TITLE
proposal: remove strings.Compare

### ISSUE THREAD

**Issue by dchenk** [golang/go#25095]
The strings.Compare function with all its documentation is:
```
// Compare returns an integer comparing two strings lexicographically.
// The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
//
// Compare is included only for symmetry with package bytes.
// It is usually clearer and always faster to use the built-in
// string comparison operators ==, <, >, and so on.
func Compare(a, b string) int {
	// NOTE(rsc): This function does NOT call the runtime cmpstring function,
	// because we do not want to provide any performance justification for
	// using strings.Compare. Basically no one should use strings.Compare.
	// As the comment above says, it is here only for symmetry with package bytes.
	// If performance is important, the compiler should be changed to recognize
	// the pattern so that all code doing three-way comparisons, not just code
	// using strings.Compare, can benefit.
	if a == b {
		return 0
	}
	if a < b {
		return -1
	}
	return +1
}
```
It's so special it even gets it's own source code file.

I suggest removing the function entirely, at least with Go 2.

I understand that it's there "only for symmetry with package bytes"—but is that really important? It's a function that has to get compiled and shipped with almost every program. (And it's sufficiently intriguing that I'll bet most Go programmers (myself included) at one point took a moment to try to understand exactly what the function even does.) The bytes and the strings packages already have a few differences.

**Comment by randall77** [comment]
It only gets shipped with a program if the program uses it.  The linker does dead code elimination of functions that are never referenced.
It won't even normally be compiled.  Standard uses of Go don't need to recompile the std lib.

I agree it is pretty useless.  But it's also pretty harmless.

**Comment by mvdan** [comment]
I agree that it's pretty useless. After four years of writing Go, I saw its only (and wrong) use only a couple of weeks ago.

But to echo Russ's and Keith's point - it's not doing any harm, and it does keep certain symmetry with the bytes package. I'm not sure if there's a benefit to removing it, given the added disadvantage of breaking backwards compatibility.

**Comment by nathany** [comment]
The existence of this function and the documentation could serve a purpose. Programmers coming from languages like Java may be looking for a function to compare strings for equality, and the comment lets them know to just use `==`.

Unfortunately, a Java programmer is more likely to end up on `bytes.Equal` (there is no `strings.Equal`) or `strings.EqualFold`. I've seen that later in the wild, and not because a case insensitive comparison was needed.

So to turn this issue around completely, adding another somewhat redundant function `strings.Equal` with a similar comment could be beneficial.

**Comment by dchenk** [comment]
@nathany the language spec describe string comparisons: https://golang.org/ref/spec#Comparison_operators

I think it'd be useful to change strings.Compare to take a []byte as the first argument and a string as the second argument. It makes sense for both arguments to be byte slices for bytes.Compare, but currently the way to compare the contents of a []byte to a string directly is to convert the string to a slice, which means an allocation—and there are instances of this in the standard library (mostly in tests). It'd be great if strings.Compare were implemented in asm just like bytes.Compare and avoided a slice allocation.
